### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by alvinwan**
goal: finish the todo list that stephen started and/or reach 90% coverage

Reminders for self:
- convert all API tests to API calls, instead of direct method invocations

Need to make sure all cases covered / Add more tests

- [x] Assignment -- permissions
- [ ] Backup -- get_messages, to_json, permissions
- [ ] Comment -- permissions
- [ ] Queue -- permissions, to_json
- [x] User -- append_email, delete_email, get_final_submission, get_or_insert, lookup
- [ ] User -- permissions
- [ ] Course -- permissions
- [ ] FinalSubmissions -- hook check (?what is this?)

Not triaged yet

- [x] FinalSubmission -- permissions
- [x] <strike>AuditLog -- none</strike>
- [ ] Group -- permissions, classmethods, invitation, acceptance, leaving, validation
- [ ] Version -- download_link, to_json, permissions
- [ ] Diff -- permissions, to_json
- [x] <strike>Score -- none</strike>
- [x] Message -- permissions (almost same as backup)
- [ ] Participant -- has_role (CM), courses (CM), permissions
- [x] Submission -- permissions (same as backup)

before branch: ~200 tests
current branch: 525 tests

### CODE REVIEW

**Comment by alvinwan** [comment]
https://github.com/coagulant/coveralls-python/issues/78

**Comment by alvinwan** [comment]
@soumyabasu @Sumukh After nearly a month, we're finally back in the green. Before this, we had ~200 tests. We now have 525, with 90.1% coverage.

I deliberately avoided testing code that we may refactor. Once you guys have reviewed/merged, I'll begin working on code cleanup.

File: server/tests/integration/test_api_group.py [hunk 1]
```diff
@@ -11,176 +11,319 @@
...
+		self.get_index(assignment=self.assignment.key.id(), member=self.user.key.id())
+		self.assertEqual(self.response_json[0]['id'], inst.key.id())
+
+		self.get_index(member=self.partner.key.id())
+		self.assertEqual(self.response_json[0]['id'], inst.key.id())
+
+	def test_remove_from_two_member_group(self):
+		self.user = self.accounts['dummy_student']
+		inst = self.model(assignment=self.assignment.key,
+											member=[self.user.key, self.accounts['dummy_student2'].key])
```
Comment (Sumukh) [inline server/tests/integration/test_api_group.py]: This line looks a little off (a lot of spacing)

File: server/tests/integration/test_api_group.py [hunk 2]
```diff
@@ -11,176 +11,319 @@
...
+			})
+
+	def test_remove_member_permissions(self):
+		""" TEsts that remove_member checks for permissions """
+		self.group = self.group.put().get()
+		with self.assertRaises(PermissionError):
+			self.API().remove_member(self.group, self.accounts['dummy_student'], {})
+
+	# def test_invite_permissions(self):
+	# 	""" Tests that invtie checks for permissions """
```
Comment (Sumukh) [inline server/tests/integration/test_api_group.py]: Why is this commented?

File: server/tests/integration/test_api_group.py (hunk 1 above)
Comment (alvinwan) [inline server/tests/integration/test_api_group.py]: Mhm, they're all tab characters though.

File: server/tests/integration/test_api_group.py (hunk 2 above)
Comment (alvinwan) [inline server/tests/integration/test_api_group.py]: A few lines down (just below this comment) on [Line 215](https://github.com/Cal-CS-61A-Staff/ok/pull/527/files#diff-49220061e3b1df2bd1b2ab33708c3694R215), I noted that the test should pass... but doesn't. Didn't want to break ok by "fixing" a non-problem... so I left the test as a reminder to investigate this.

**Comment by Sumukh** [comment]
This is great! Awesome work @alvinwan 

` coverage/coveralls — Coverage increased (+53.0%) to 90.141%` :open_mouth:  :tada: :trophy: 

 I did a quick first pass - but looking at this in detail might take a while.

**Comment by alvinwan** [comment]
Thanks Sumukh! Haha, yeah - there's all of 3k+ lines to peruse. Feel free to change the tabs back to spaces. I was just resolving a merge conflict.

File: server/tests/integration/test_api_group.py (hunk 2 above)
Comment (alvinwan) [inline server/tests/integration/test_api_group.py]: Fixed. The test is uncommented in the newer #565 PR... Turns out someone wrote `return need.exception()` instead of `raise need.exception()`, so I fixed that.

File: server/app/api.py [hunk 3]
```diff
@@ -50,20 +71,9 @@ def parse_json_field(field):
...
-    :return: (string) parsed JSON
-    """
-    if not field[0] == '[':
-        if field == 'false':
-            return False
-        elif field == 'true':
-            return True
-        return field
-    return json.loads(field)
+parse_json_list_field = parse_json_field
```
Comment (Sumukh) [inline server/app/api.py]: I'm guessing there was some reason for this separation - but merging it into one function seems ok.

File: server/app/api.py [hunk 4]
```diff
@@ -674,9 +684,6 @@ def get_backups(self, obj, user, data):
     def get_submissions(self, obj, user, data):
         return [subm.submission for subm in obj.get_submissions(data['assignment'], data['quantity'])]
 
-    def timed_submission(self, obj, user, data):
-        return obj.get_submission_before(data['assignment'], data['before'])
```
Comment (Sumukh) [inline server/app/api.py]: @alvinwan - Why was this removed? I have a feeling we used it for a script and might need it - but I could be wrong.

**Comment by Sumukh** [comment]
I did another pass through (though I skimmed over the tests) and it looks fine. I'm ok with merging this - @alvinwan - go ahead and merge if this ready. 

Coverage is down to 89.877% after merging with master :disappointed:

File: server/app/api.py (hunk 3 above)
Comment (alvinwan) [inline server/app/api.py]: Might've been. Yeah, nothing broke.

File: server/app/api.py (hunk 4 above)
Comment (alvinwan) [inline server/app/api.py]: Er, didn't find it in the codea nd it wasn't exposed as an endpoint. Unsure though.
